modules/002-deckhouse/hooks/check_kernel_versions.go

```diff
@@ -0,0 +1,150 @@
+	KernelVersion string
+	SemverVersion *semver.Version
+}
+
+type nodeConstraint struct {
+	KernelVersionConstraint string
+	ModulesListInUse        []string
+}
+
+var constraints = []nodeConstraint{
+	{
+		KernelVersionConstraint: ">= 4.9.17",
+		ModulesListInUse:        []string{"cni-cilium"},
+	},
+	{
+		KernelVersionConstraint: ">= 5.7",
+		ModulesListInUse:        []string{"cni-cilium", "istio"},
+	},
+	{
+		KernelVersionConstraint: ">= 5.7",
+		ModulesListInUse:        []string{"cni-cilium", "openvpn"},
+	},
+}
+
+const (
+	nodeKernelCheckMetricsGroup = "node_kernel_check"
+	nodeKernelCheckMetricName   = "d8_node_kernel_does_not_satisfy_requirements"
+)
+
+var _ = sdk.RegisterFunc(&go_hook.HookConfig{
+	Kubernetes: []go_hook.KubernetesConfig{
+		{
+			Name:       "nodes",
+			ApiVersion: "v1",
+			Kind:       "Node",
+			LabelSelector: &metav1.LabelSelector{
+				MatchExpressions: []metav1.LabelSelectorRequirement{
+					{
+						Key:      "node.deckhouse.io/group",
+						Operator: metav1.LabelSelectorOpExists,
+					},
+				},
+			},
+			FilterFunc: filterNodes,
+		},
+	},
+}, handleNodes)
+
+func filterNodes(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var node corev1.Node
+
+	err := sdk.FromUnstructured(obj, &node)
+	if err != nil {
+		return nil, err
+	}
+
+	v, err := semver.NewVersion(strings.Split(node.Status.NodeInfo.KernelVersion, "-")[0])
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse kernel version %s for node %s: %v", node.Status.NodeInfo.KernelVersion, node.Name, err)
+	}
+
+	return nodeKernelVersion{
+		Name:          node.Name,
+		KernelVersion: node.Status.NodeInfo.KernelVersion,
+		SemverVersion: v,
+	}, nil
+}
+
+func handleNodes(input *go_hook.HookInput) error {
+	input.MetricsCollector.Expire(nodeKernelCheckMetricsGroup)
+
+	snap := input.Snapshots["nodes"]
+	if len(snap) == 0 {
+		return nil
+	}
+
+	enabledModules := set.NewFromValues(input.Values, "global.enabledModules")
+
+	for _, constrant := range constraints {
+		// check modules in use
+		check := true
+		for _, m := range constrant.ModulesListInUse {
+			if !enabledModules.Has(m) {
+				check = false
+				break
+			}
+		}
+		if !check {
+			continue
+		}
+
+		c, err := semver.NewConstraint(constrant.KernelVersionConstraint)
+		if err != nil {
+			return err
+		}
+
+		for _, n := range snap {
+			if n == nil {
+				continue
+			}
+			node := n.(nodeKernelVersion)
+			if !c.Check(node.SemverVersion) {
+				input.MetricsCollector.Set(nodeKernelCheckMetricName, 1, map[string]string{"node": node.Name, "kernel_version": node.KernelVersion, "affected_module": strings.Join(constrant.ModulesListInUse, ","), "constraint": constrant.KernelVersionConstraint}, metrics.WithGroup(nodeKernelCheckMetricsGroup))
+				input.LogEntry.Errorf("kernel %s on node %s does not satisfy kernel constraint %s for modules [%s]", node.KernelVersion, node.Name, constrant.KernelVersionConstraint, strings.Join(constrant.ModulesListInUse, ","))
```

Comment:
This hook reacts to changes in any node object. If there is a problem in the cluster with 50 nodes and autoscaling, we will see 50 error messages in the log each time a node is added or deleted.

I suggest avoiding it and logging only cumulative messages / do not logging it at all because we have metrics.

Reply:
I added verbose debug logging , and cumulative message, as we discussed.